src/transports/dash/integrity_checks.ts

```diff
@@ -0,0 +1,186 @@
+          try {
+            trowOnIntegrityError(data);
+            callbacks.onNewChunk(data);
+          } catch (err) {
+            // Do not reject with a `CancellationError` after cancelling the request
+            cleanUpCancellers();
+
+            // Cancel the request
+            requestCanceller.cancel();
+
+            // Reject with thrown error
+            reject(err);
+          }
+        },
+      }).then(
+        (info) => {
+          cleanUpCancellers();
+          if (requestCanceller.isUsed()) {
+            return;
+          }
+          if (info.resultType === "segment-loaded") {
+            try {
+              trowOnIntegrityError(info.resultData.responseData);
+            } catch (err) {
+              reject(err);
+              return;
+            }
+          }
+          resolve(info);
+        },
+        (err: unknown) => {
+          cleanUpCancellers();
+          reject(err);
+        },
+      );
+
+      function cleanUpCancellers() {
+        requestCanceller.signal.deregister(reject);
+        unlinkCanceller();
+      }
+    });
+
+    /**
+     * If the data's seems to be corrupted, throws an `INTEGRITY_ERROR` error.
+     * @param {*} data
+     */
+    function trowOnIntegrityError(data: T): void {
+      if (
+        (!(data instanceof ArrayBuffer) && !(data instanceof Uint8Array)) ||
+        inferSegmentContainer(context.type, context.mimeType) !== "mp4"
+      ) {
+        return;
+      }
+      checkISOBMFFIntegrity(new Uint8Array(data), context.segment.isInit);
+    }
+  };
+}
+
+/**
+ * Add multiple checks on the response given by the `manifestLoader` in argument.
+ * If the response appear to be corrupted, the returned Promise will reject with
+ * an error with an `INTEGRITY_ERROR` code.
+ * @param {Function} manifestLoader
+ * @returns {Function}
+ */
+export function addManifestIntegrityChecks(
+  manifestLoader: ITransportManifestPipeline["loadManifest"],
+): ITransportManifestPipeline["loadManifest"] {
+  return async (url, options, initialCancelSignal) => {
+    const res = await manifestLoader(url, options, initialCancelSignal);
+    trowOnIntegrityError(res.responseData);
+    return res;
+
+    /**
+     * If the data's seems to be corrupted, throws an `INTEGRITY_ERROR` error.
+     * @param {*} data
+     */
+    function trowOnIntegrityError(data: ILoadedManifestFormat): void {
+      if (typeof data === "string") {
+        let currOffset = data.length - 1;
+        while (isCharXmlWhiteSpace(data[currOffset])) {
+          currOffset--;
+        }
+        if (data[currOffset] !== ">") {
+          throw new OtherError("INTEGRITY_ERROR", "MPD does not end with </MPD>");
+        }
+        currOffset--;
+        while (isCharXmlWhiteSpace(data[currOffset])) {
+          currOffset--;
+        }
+        if (data.substring(currOffset - 2, currOffset + 1) !== "MPD") {
+          throw new OtherError("INTEGRITY_ERROR", "MPD does not end with </MPD>");
+        }
```

Comment:
Another possible implementation could be to iterate through the expected string to be found, not sure which one is the best for this use case

```js
  let buffer = ["<", "MPD", "/", ">"];

  let currOffset = data.length - 1;
  for (let i = buffer.length; i > 0; i--) {
    let currentWord = buffer[i - 1];
    while (isCharXmlWhiteSpace(data[currOffset])) {
      currOffset--;
    }
    for (let j = currentWord.length; j > 0; j--) {
      if (data[currOffset] !== currentWord[j - 1]) {
        throw new Error("INTEGRITY_ERROR MPD does not end with </MPD>");
      } else {
        currOffset--;
      }
    }
  }

```

Reply:
Yeah your code seems simpler to me, will update.

Reply:
Done